src/attribute-based/ABE.c

```diff
@@ -0,0 +1,736 @@
+                mpz_clear(sum);
+            }
+        }
+
+        destroyPolynom(q);
+    }
+    else
+    {
+        AffinePoint Cy;
+        CryptidStatus status = affine_wNAFMultiply(&Cy, publickey->g, s, publickey->ellipticCurve);
+        if(status) {
+            affine_destroy(Cy);
+            return status;
+        }
+
+        // H(att(x))
+        AffinePoint hashedPoint;
+
+        status = hashToPoint(&hashedPoint, accessTree->attribute, accessTree->attributeLength, publickey->q, publickey->ellipticCurve, publickey->hashFunction);
+
+        if (status) 
+        {
+            return status;
+        }
+
+        AffinePoint CyA;
+        status = affine_wNAFMultiply(&CyA, hashedPoint, s, publickey->ellipticCurve);
+        if(status) {
+            return status;
+        }
+
+        affine_destroy(hashedPoint);
+
+        accessTree->Cy = Cy;
+        accessTree->CyA = CyA;
+        accessTree->computed = 1;
+    }
+
+    return CRYPTID_SUCCESS;
+}
+
+// Encrypts message with the specified accessTree and publicKey to encrypted
+CryptidStatus cryptid_encrypt_ABE(EncryptedMessage_ABE* encrypted,
+                                const char *const message, const size_t messageLength,
+                                    PublicKey_ABE* publickey, AccessTree* accessTree)
+{
+    if(!message)
+    {
+        return CRYPTID_MESSAGE_NULL_ERROR;
+    }
+
+    if(messageLength == 0)
+    {
+        return CRYPTID_MESSAGE_LENGTH_ERROR;
+    }
+
+    if(!publickey) {
+        return CRYPTID_MESSAGE_NULL_ERROR;
+    }
+
+    if(!accessTree) {
+        return CRYPTID_MESSAGE_NULL_ERROR;
+    }
+
+    mpz_t pMinusOne;
+    mpz_init(pMinusOne);
+    mpz_sub_ui(pMinusOne, publickey->ellipticCurve.fieldOrder, 1);
+
+    mpz_t s;
+    mpz_init(s);
+    random_mpzInRange(s, pMinusOne);
+    compute_tree(accessTree, s, publickey);
+
+    encrypted->tree = accessTree;
+    Complex eggalphas;
+    complex_modPow(&eggalphas, publickey->eggalpha, s, publickey->ellipticCurve.fieldOrder);
+
+    mpz_t M;
+    mpz_init(M);
+    size_t n = messageLength;
+    size_t startFrom = 0;
+    char* msg = malloc(sizeof(char)*messageLength);
+    strncpy(msg, message, messageLength);
+    Ctilde_set* prevSet = (Ctilde_set*) malloc(sizeof(Ctilde_set));
+    encrypted->Ctilde_set = prevSet;
+    prevSet->last = 2;
+
+    // Splitting message to parts if M >= ellipticCurve.fieldOrder
+    // Splitted parts of the message is stored in a set (Ctilde_set)
+    while(n > 0)
+    {
+        while(prevSet->last == 2 || mpz_cmp(M, publickey->ellipticCurve.fieldOrder) >= 0)
+        {
+            if(prevSet->last == 0) {
+                n--;
+            }
+            prevSet->last = 0;
+            char* msgPart = malloc(sizeof(char)*(n+1));
+            strncpy(msgPart, msg + 0, n);
+            msgPart[n] = '\0';
+            mpz_import (M, n+1, 1, 1, 0, 0, msgPart);
+            free(msgPart);
+        }
```

Comment:
It is needed, value of M is calculated the following way:
`mpz_import (M, n+1, 1, 1, 0, 0, msgPart);`
The value of n is subtracted inside the inner while loop until the value of M is less than the fieldOrder. (It would be possible to calculate n without a loop, but that would not be that perfect as this solution from the aspect of maximalizing lengths of splitted parts.)

Reply:
> tmp_outer: 1, tmp_inner: 1
tmp_outer: 1, tmp_inner: 2
tmp_outer: 1, tmp_inner: 3
tmp_outer: 1, tmp_inner: 4
tmp_outer: 1, tmp_inner: 5
tmp_outer: 1, tmp_inner: 6
tmp_outer: 1, tmp_inner: 7
tmp_outer: 1, tmp_inner: 8
tmp_outer: 1, tmp_inner: 9
tmp_outer: 1, tmp_inner: 10
tmp_outer: 1, tmp_inner: 11
...
tmp_outer: 1, tmp_inner: 417

To make it sure, I have added debug messages and iteration counter integers to the code.